File: Sheets/src/neetcode/practice/Binary_Search/_1_binarySearch.java
package neetcode.practice.Binary_Search;

public class _1_binarySearch {
  public static int search(int[] nums, int target) {
    int n = nums.length;

    int l = 0, r = n - 1;

    while (l <= r) {
      int m = (l + r) / 2;
      if (nums[m] < target) l = m + 1;
      else if (nums[m] > target) r = m - 1;
      else return m;
    }

    return -1;
  }
}

File: Sheets/src/neetcode/practice/Binary_Search/_1_BinarySearch.cc
#include <vector>

class Solution {
public:
  int search(std::vector<int> &nums, int target) {
    ptrdiff_t i = 0, j = nums.size() - 1;

    while (i <= j) {
      ptrdiff_t m = (i + j) >> 1;
      if (target > nums[m])
        i = m + 1;
      else if (target < nums[m])
        j = m - 1;
      else
        return static_cast<int>(m);
    }

    return -1;
  }
};
